[🦩] Add swiftASTGen library to build a C++ AST from a SwiftSyntax parse tree

DougGregor opened the pull request:
Introduce a `swiftASTGen` library, written as a Swift module, which produces a C++ AST from the corresponding SwiftSyntax syntax tree as produced by the new Swift parser. Architecturally, this new library sits between the C++ AST library (since it constructs ASTs) and the C++ parser (whose entry points will be diverted through this library). This new code path intercepts the main entry point of the C++ parser, which parses "top-level" declarations, when the command-line option `--enable-experimental-feature ParserASTGen`. It doesn't do much yet; this pull request is primarily to get all of the major architectural pieces in place so we can iterate on them.

`swiftASTGen` depends on the `CASTBridging` layer, a painstakingly-crafted C API that is used to build and inspect C++ AST nodes, so they can be used from Swift. We cannot currently depend on Swift/C++ interoperability, because it is an experimental feature, whereas this library will need to be able to build with a host Swift compiler from up to a year ago. 

The CMake build system is the only one that's able to produce a working compiler. There is also a `Package.swift` that makes it easier to build and work with the `swiftASTGen` library within IDEs, but it's mainly there for editing---it won't create something that can be meaningfully executed. Most things with the new Swift parser are better implemented/tested within or on top of the `swift-syntax` package.

zoecarver committed:
Add Swift-side infrastructure for ASTGen.

(cherry picked from commit 7bc67c9f12d3ae4f4fc2c2151d39e4ea039a58ff)

zoecarver committed:
Move ASTGen into Swift from swift-syntax.

(cherry picked from commit 2df25fc6e0dd7d63a7f5c5a079b92db0cae87a3a)

zoecarver committed:
Apply Doug's review comments

(cherry picked from commit ddede209c72e532e39864dca63bbdbb97362665f)

zoecarver committed:
Make ASTGen a swift package

(cherry picked from commit 4e9e5fe6a83d155b269e21ee7410f758cca14004)

zoecarver committed:
Add support for structs/classes.

(cherry picked from commit 1b1944ea4840cd8b4b01921501c09f6db3e57fcf)

zoecarver committed:
[nfc] Split ASTGen into files + extensions.

(cherry picked from commit 629fb688f07b58ba50f5ce8f077da903dad2e073)

zoecarver committed:
Add support for bool literals.

(cherry picked from commit d58a01db958a3bdad1648f7d4d96d56793d912e4)

zoecarver committed:
Support functions with statements in their body.

(cherry picked from commit 5935670931b0120a2383d2a9c51e8af4c7ad58dd)

DougGregor committed:
Remove unnecessary frontend option

DougGregor committed:
Rename experimental feature to ParserASTGen

DougGregor committed:
Clean up build logic for ASTGen library.

Only introduce it and its dependency when the new Swift parser is being
built, and rely more on existing logic to make sure we get the right
build/link flags.

DougGregor committed:
Use ASTGen with the new Swift parser when requested

When the experimental flag `ParserASTGen` is enabled, the compiler has
ASTGen built, we're not performing code completion, and we're not
parsing SIL... go through the new parser and ASTGen to build the
abstract syntax tree.

DougGregor committed:
Fix the package that builds swiftASTGen.

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please test source compatibility

zoecarver commented:
Thank you for fixing this up, Doug!

DougGregor committed:
Use add_library for Swift ASTGen library

DougGregor committed:
Fix CASTBridging.h for non-Clang host compilers

DougGregor commented:
@swift-ci please smoke test

etcwilde reviewed (commented):
Lets remove the `xcworkspacedata` file and Zoe's `UserInterfaceState.xcuserstate`.

krzyzanowskim commented on lib/ASTGen/Sources/ASTGen/ASTGen.swift:
```diff
@@ -0,0 +1,78 @@
+}
+
+// TODO: :(
+var declContext: UnsafeMutableRawPointer! = nil
```
mutable module-wide global variable buried between the lines.

It feels like it supposes to be a `ASTGenVisitor` property set in the `ASTGenVisitor` initializer. Unfortunate `ASTGenVisitor` is a struct that makes it all `mutating`. How about wrapping it in a class

```swift
class DeclContext {
    var declContext: UnsafeMutableRawPointer! = nil
}

struct ASTGenVisitor ... {
    var declContext = DeclContext()
}

```

and yes, nobody asked me for this comment

DougGregor replied:
Gaaa, I hadn't noticed that. I've done a Sneaky Property Wrapper Trick for the moment.

DougGregor committed:
[CMake] Make sure swiftASTGen links swiftAST.

DougGregor committed:
[ASTGen] Use UTF-8 rather than C `char`, and use BridgedIdentifier consistently

DougGregor committed:
Don't try to bridge diagnostics from C++ to Swift

We'll be going the other way, with diagnostics existentials.

DougGregor committed:
[ASTGen] Clean up casting with source locations

DougGregor committed:
[ASTGen] Make sure we spin the C++ parser even when we don't use it.

This is a temporary hack; we shouldn't even need to create the C++
parser here.

DougGregor committed:
Remove xcworkspacedata and UserInterfaceState.xcuserstate

DougGregor committed:
[ASTGen] Do a little property-wrapper dance to move DeclContext into the visitor

DougGregor commented:
@swift-ci please test

DougGregor commented:
This *really* doesn't do much now, and lots of stuff is likely to get reworked. We need to do a deep dive on a small part of the grammar (types and `typealias` is probably the smallest bit) and make things as clean and complete as possible, and see how that goes.

ahoppen commented on lib/AST/DeclContext.cpp:
```diff
@@ -882,7 +882,7 @@ void IterableDeclContext::addMember(Decl *member, Decl *hint, bool insertAtHead)
 void IterableDeclContext::addMemberSilently(Decl *member, Decl *hint,
                                             bool insertAtHead) const {
-  assert(!isa<AccessorDecl>(member) && "Accessors should not be added here");
+//  assert(!isa<AccessorDecl>(member) && "Accessors should not be added here");
```
Is the removal of this assertion intentional?

DougGregor replied:
No, it is not! I'll bring it back.

ahoppen commented on lib/ASTGen/Package.swift:
```diff
@@ -0,0 +1,32 @@
+
+import PackageDescription
+
+let package = Package(
```
Would it make sense to add this paragraph of yours to Package.swift?
> The CMake build system is the only one that's able to produce a working compiler. There is also a Package.swift that makes it easier to build and work with the swiftASTGen library within IDEs, but it's mainly there for editing---it won't create something that can be meaningfully executed. Most things with the new Swift parser are better implemented/tested within or on top of the swift-syntax package.

DougGregor replied:
Great idea!

ahoppen reviewed (approved):
The overall build structure looks good to me. I didn’t review any of the new Swift or C++ code because IIUC that’s not the intention of this PR.

DougGregor committed:
[ASTGen] Fix integer and string literal mapping

DougGregor committed:
Bring back an important assertion

DougGregor committed:
[ASTGen] Add a comment about the purporse of its Package.swift

DougGregor committed:
[CMake] Only add swiftCore dependency when the new Swift parser is included.

DougGregor committed:
[CMake] Ensure that the SourceKit framework links Swift as needed

DougGregor committed:
[CMake Export swiftASTGen so the LLDB target can use it

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
TO BE CLEAR, I have no idea what the flamingo is about. @zoecarver gonna hafta answer that one. I just work here.

zoecarver commented:
> TO BE CLEAR, I have no idea what the flamingo is about. @zoecarver gonna hafta answer that one. I just work here.

Just keeping things organized...

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test macOS

DougGregor commented:
Windows test failure is unrelated

DougGregor committed:
[ASTGen] Set the target triple for swiftASTGen.

Builds of Swift sources via CMake are not getting a target triple set,
so they pick up the host triple. This leads to spurious warnings about
linking in code built for a newer OS version (e.g., the one you're on)
into libraries with an older deployment target. It also breaks
building a cross-compiler. Always set the target triple.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please build toolchain
